doc/dlt-adaptor-udp.1.md
% DLT-ADAPTOR-UDP(1)

# NAME

**dlt-adaptor-udp** - Forward received UDP messages to DLT Daemon

# SYNOPSIS

**dlt-adaptor-udp** \[**-a** apid\] \[**-c** ctid\] \[**-p**\] \[**-h**\]

# DESCRIPTION

This is a small external program for forwarding received UDP messages to DLT
Daemon.

This also can be used for e.g. sending syslog messages to the DLT daemon.
Therefore a syslog daemon called *syslog-ng* is necessary. This syslog daemon
must be configured to send the syslog messages to a specific UDP port. For
configuration of this syslog daemon, see the documentation for *syslog-ng*.
This tools is already integrated into *dlt-system*.

## OPTIONS

-a

:    Set application ID to apid (default: UDPA)

-c

:    Set context ID tp ctid (default: UDPC)

-p

:    Set receive port number for UDP messages (default: 47111)

-h

:    Show help

# EXIT STATUS

Non zero is returned in case of failure.

# AUTHOR

Saya Sugiura (ssugiura (at) jp.adit-jv (dot) com)

# COPYRIGHT

Copyright (C) 2019 Advanced Driver Information Technology, Bosch and DENSO. License MPL-2.0: Mozilla Public License version 2.0 http://mozilla.org/MPL/2.0/.

# BUGS

See Github issue: <https://github.com/COVESA/dlt-daemon/issues>

# SEE ALSO

**dlt-daemon(1)**


ReleaseNotes.md
# Diagnostic Log and Trace  - Release Notes

Back to [README.md](../README.md)

## Version

2.18.9 STABLE

## Changes

### 2.18.9

   * dlt-user: fix crash with certain strings (#463)
   * dlt_multiple_files: remove superfluous mode bits and add header file to header list (#462)
   * Android: Add new feature in Android bp (#461)
   * cmake: remove duplicated option message (#454)
   * house-keeper: remove infinite wait (#438)
   * dlt-logd-converter: Fix getting log level from log msg (#456)
   * dlt-logd-converter: fixes android 12 compilation (#445)
   * logfile: exhance internal dlt logging by introducing size limits (#369)
   * This changes a mispatch from fcb676a7 to install the udp binary correctly. (#449)
   * Installs dlt.conf on android (#446)
   * dlt-connection: add socket timeout (#439)
   * Fix memory leak (#441)
   * Check for negative index in dlt_file_message (#437)
   * dlt-user: fix potential non closed socket in init/free (#435)
   * dlt-convert: Fix memory leak by calling dlt_file_free (#434)
   * dlt-user: Fix crashes in dlt_free during dlt_init (#362)
   * Update contacts and removed mailing lists (#431)
   * Updates for Coding Styles (#425)
   * gateway: Fix Node handling and ECUid checks (#429)
   * filetransfer: fix filesize divisible by blocksize case (#383)
   * client: Fix Get Log Info response conversion method (#422)
   * cmake: network trace enable toggle (#424)
   * dlt-system: Fix buffer overflow detection on 32bit targets (#398)
   * dlt-receive: set host interface and allow multiple udp multicast addresses (#420)
   * Fix for Resource and Memory Leak (#418)
   * dlt_daemon_client: Fix Control Msg ECUId comparison with active Gateway (#414)
   * Avoid memory corruption behind buffer wp in function dlt_getloginfo_conv_ascii_to_id (#411)
   * dlt_common: change output of message for log initialization (#412)
   * internal-logging: Fix issues with file logging (#378)
   * systemd: add support for socket activation via systemd (#401)
   * Update maintainer (#410)
   * dlt_daemon_client: Fix change loglevel of application (#408)
   * dlt_client:Block in connect() (#409)
   * dlt-gateway: Fix crash on invalid ip (#381)
   * Update dlt_for_developers.md (#405)
   * logstorage: Truncate ECUid in Logstorage filter to prevent crash (#402)
   * dlt_common.c: Change default logging_mode (#406)
   * dlt-daemon-connection: Start up even if not all bindings are valid (#380)
   * enforce-trace-limit: ContextLogLevel is now enforced in the daemon (#382)
   * automotive-dlt.pc: add the path to find the static library (#387)
   * systemd: install adaptor-udp service for adaptor=on (#393)
   * Fix handle returned value (#384)
   * README: Update link to github actions (#392)
   * Update for CI (#389)
   * Fix a double-free bug. (#376)
   * Issue-ID: make-adaptor-configurablecmake: Add option to enable each adaptor by itself (#364)
   * Fix the target name in documentation (#372)
   * cmake: Add options to enable/disable each dlt console tool (#363)
   * filetransfer: Fix getFileCreationDate2 stat check (#361)
   * tests: Deplicate unused files and variables (#359)
   * Fix DLT User/Client tests (#357)
   * lib: Correct VARI usage in dlt_user_log_write_uint (#356)
   * filetransfer: Return error if no free space (#354)
   * Support for Cygwin toolchain. (#351)
   * dlt-system: fix invalid free by removing unused TempDir (#350)
   * fix -Wformat issues reported by clang (#349)
   * Forcibly the severity level set (#346)
   * daemon: Do not exit when accept returns ECONNABORTED (#347)
   * dlt-system : fix invalid free with ConfigurationFileName (#342)
   * dlt-daemon: Only create directories if they do not exist yet (#340)
   * fixes compilation issue with clang (#339)
   * dlt-daemon: create sockets using "android way" (#333)
   * dlt-system: fix a libc buffer overflow detection on 32bit targets (#337)

### 2.18.8

   * lib: Fix wrong type alert from lgtm
   * gtest_dlt_daemon_gateway: fix gtest build failed
   * lib: generate dlt library internal log file
   * tests: add stdlib to dlt_cpp_extension
   * dlt_user: Make dlt_init thread safe
   * remove clang-tidy analyzer warnings: incompatible pointer type
   * debian: improve debian build package
   * dlt-control-common: shutdown and close socket
   * dlt_common: improve function description
   * gtest: Bring-in changes
   * gtest: Rework WORKING_DIRECTORY
   * gtest: Correct data amount in gtest_dlt_common
   * cmake: Set empty to systemd_SRCS
   * gtest: Refactor tests/CMakeLists.txt
   * lib: Add SOCK_CLOEXEC to socket
   * daemon: Create parent directory for unix socket
   * cmake: Correct added subdirectories
   * console: Add cmake options for control and timestamp
   * tests: Adapt to DLT_DISABLE_MACRO
   * header: Adapt to DLT_DISABLE_MACRO
   * include: Refactor CMakeLists.txt
   * gtest: Change script name to gtest_dlt_daemon_offline_log.sh
   * daemon: Enable to use FIFO on QNX
   * tests: Add new test case with given buffer
   * tests: Enable macro disabling
   * lib: Add new interfaces with given buffer
   * Implemention of tests for the dlt-qnx-system module
   * lib: Add MaxFileSize handling
   * client: pthread_join for deinit
   * doc: update initial log level document
   * dlt-system: Fix memory leak in dlt-system config
   * dlt_common: remove duplicate stdbool header
   * dlt-control: Add option to config port
   * system: use signalfd for dlt-system
   * console: provides args option to enable send/receive serial header
   * fix malformed printf format strings (#295)
   * cmake: Set WITH_LEGACY_INCLUDE_PATH to ON as default (#334)
   * Make the legacy include path a CMake option (#332)
   * daemon: Call dlt_daemon_configuration_load() properly (#330)
   * dlt_user: Use pthread_setname_np() if available (#326)
   * libdlt: Add legacy include path in exported CMake config file (#327)
   * lib: Set TYLE to 1 for BOOL type (#320)
   * file-transfer: Abort file transfer if get serial number failed

src/adaptor/dlt-adaptor-udp.c
/*
 * SPDX license identifier: MPL-2.0
 *
 * Copyright (C) 2011-2015, BMW AG
 *
 * This file is part of COVESA Project DLT - Diagnostic Log and Trace.
 *
 * This Source Code Form is subject to the terms of the
 * Mozilla Public License (MPL), v. 2.0.
 * If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * For further information see http://www.covesa.org/.
 */

/*!
 * \author Alexander Wenzel <alexander.aw.wenzel@bmw.de>
 *
 * \copyright Copyright © 2011-2015 BMW AG. \n
 * License MPL-2.0: Mozilla Public License version 2.0 http://mozilla.org/MPL/2.0/.
 *
 * \file dlt-adaptor-udp.c
 */

/*******************************************************************************
**                                                                            **
**  SRC-MODULE: dlt-adaptor-udp.c                                             **
**                                                                            **
**  TARGET    : linux                                                         **
**                                                                            **
**  PROJECT   : DLT                                                           **
**                                                                            **
**  AUTHOR    : Alexander Wenzel Alexander.AW.Wenzel@bmw.de                   **
**              Markus Klein                                                  **
**                                                                            **
**  PURPOSE   :                                                               **
**                                                                            **
**  REMARKS   :                                                               **
**                                                                            **
**  PLATFORM DEPENDANT [yes/no]: yes                                          **
**                                                                            **
**  TO BE CHANGED BY USER [yes/no]: no                                        **
**                                                                            **
*******************************************************************************/

/*******************************************************************************
**                      Author Identity                                       **
********************************************************************************
**                                                                            **
** Initials     Name                       Company                            **
** --------     -------------------------  ---------------------------------- **
**  aw          Alexander Wenzel           BMW                                **
**  mk          Markus Klein               Fraunhofer ESK                     **
*******************************************************************************/

/*******************************************************************************
**                      Revision Control History                              **
*******************************************************************************/

/*
 * $LastChangedRevision: 1670 $
 * $LastChangedDate: 2011-04-08 15:12:06 +0200 (Fr, 08. Apr 2011) $
 * $LastChangedBy$
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <errno.h>

#include "dlt_common.h"
#include "dlt_user.h"
#include "dlt_user_macros.h"


/* Port number, to which the syslogd-ng sends its log messages */
#define RCVPORT               47111

#define MAXSTRLEN             1024

#define PU_DLT_APP_DESC      "udp adaptor application"
#define PU_DLT_CONTEXT_DESC  "udp adaptor context"

#define PU_DLT_APP "UDPA"
#define PU_DLT_CONTEXT "UDPC"

DLT_DECLARE_CONTEXT(mycontext)

int main(int argc, char *argv[])
{
    int sock;
    int bytes_read;
    socklen_t addr_len;
    int opt, port;
    char recv_data[MAXSTRLEN];
    struct sockaddr_in client_addr, server_addr;

    char apid[DLT_ID_SIZE];
    char ctid[DLT_ID_SIZE];
    char version[255];
    int verbosity = DLT_LOG_INFO;

    dlt_set_id(apid, PU_DLT_APP);
    dlt_set_id(ctid, PU_DLT_CONTEXT);

    port = RCVPORT;

    while ((opt = getopt(argc, argv, "a:c:hp:v:")) != -1)
        switch (opt) {
        case 'a':
        {
            dlt_set_id(apid, optarg);
            break;
        }
        case 'c':
        {
            dlt_set_id(ctid, optarg);
            break;
        }
        case 'h':
        {
            dlt_get_version(version, 255);

            printf("Usage: dlt-adaptor-udp [options]\n");
            printf("Adaptor for forwarding received UDP messages to DLT daemon.\n");
            printf("%s \n", version);
            printf("Options:\n");
            printf("-a apid      - Set application id to apid (default: UDPA)\n");
            printf("-c ctid      - Set context id to ctid (default: UDPC)\n");
            printf("-p           - Set receive port number for UDP messages (default: %d) \n", port);
            printf(
                "-v verbosity level - Set verbosity level (Default: INFO, values: FATAL ERROR WARN INFO DEBUG VERBOSE)\n");
            printf("-h           - This help\n");
            return 0;
            break;
        }
        case 'p':
        {
            port = atoi(optarg);
            break;
        }
        case 'v':
        {
            if (!strcmp(optarg, "FATAL")) {
                verbosity = DLT_LOG_FATAL;
                break;
            }
            else if (!strcmp(optarg, "ERROR"))
            {
                verbosity = DLT_LOG_ERROR;
                break;
            }
            else if (!strcmp(optarg, "WARN"))
            {
                verbosity = DLT_LOG_WARN;
                break;
            }
            else if (!strcmp(optarg, "INFO"))
            {
                verbosity = DLT_LOG_INFO;
                break;
            }
            else if (!strcmp(optarg, "DEBUG"))
            {
                verbosity = DLT_LOG_DEBUG;
                break;
            }
            else if (!strcmp(optarg, "VERBOSE"))
            {
                verbosity = DLT_LOG_VERBOSE;
                break;
            }
            else {
                printf(
                    "Wrong verbosity level, setting to INFO. Accepted values are: FATAL ERROR WARN INFO DEBUG VERBOSE\n");
                verbosity = DLT_LOG_INFO;
                break;
            }

            break;
        }
        default: /* '?' */
        {
            fprintf(stderr, "Unknown option '%c'\n", optopt);
            exit(3);
            return 3;/*for parasoft */
        }
        }


#ifdef DLT_USE_IPv6

    if ((sock = socket(AF_INET6, SOCK_DGRAM, 0)) == -1)
#else

    if ((sock = socket(AF_INET, SOCK_DGRAM, 0)) == -1)
#endif
    {
        perror("Socket");
        exit(1);
    }

#ifdef DLT_USE_IPv6
    server_addr.sin_family = AF_INET6;
#else
    server_addr.sin_family = AF_INET;
#endif
    server_addr.sin_port = htons(port);
    server_addr.sin_addr.s_addr = INADDR_ANY;
    memset(&(server_addr.sin_zero), 0, 8);
    if (bind(sock, (struct sockaddr *)&server_addr,
             sizeof(struct sockaddr)) == -1) {
        perror("Bind");
        return -1;
    }

    addr_len = sizeof(struct sockaddr);

    DLT_REGISTER_APP(apid, PU_DLT_APP_DESC);
    DLT_REGISTER_CONTEXT(mycontext, ctid, PU_DLT_CONTEXT_DESC);

    while (1) {
        bytes_read = 0;

        bytes_read = recvfrom(sock, recv_data, MAXSTRLEN, 0,
                              (struct sockaddr *)&client_addr, &addr_len);

        if (bytes_read == -1) {
            if (errno == EINTR) {
                continue;
            }
            else {
                DLT_UNREGISTER_CONTEXT(mycontext);
                DLT_UNREGISTER_APP();
                exit(1);
            }
        }

        recv_data[bytes_read] = '\0';

        if (bytes_read != 0)
            DLT_LOG(mycontext, verbosity, DLT_STRING(recv_data));
    }

    DLT_UNREGISTER_CONTEXT(mycontext);
    DLT_UNREGISTER_APP();

    return 0;
}
